### PR TITLE
Treat IPeekResultFactory as optional import

### PR DESCRIPTION
**Pull request by KirillOsenkov**
It is not available on Visual Studio for Mac so we want to gracefully degrade when it's not around.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Lgtm

**Comment by CyrusNajmabadi** [comment]
I am logged in as me... But I'm missing the review button...

File: src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs [hunk 1]
```diff
@@ -24,8 +24,8 @@ internal sealed class PeekableItemSourceProvider : IPeekableItemSourceProvider
 
         [ImportingConstructor]
         public PeekableItemSourceProvider(
-            IPeekableItemFactory peekableItemFactory,
-            IPeekResultFactory peekResultFactory,
+            [Import(AllowDefault = true)] IPeekableItemFactory peekableItemFactory,
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs]: 📝 This is a required export. The implementation is defined and exported in the same assembly where the interface is defined.

File: src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs [hunk 2]
```diff
@@ -24,8 +24,8 @@ internal sealed class PeekableItemSourceProvider : IPeekableItemSourceProvider
...
+            [Import(AllowDefault = true)] IPeekResultFactory peekResultFactory,
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs]: 📝 This is intentionally not optional. By design, `PeekableItemSourceProvider` will be silently omitted from the catalog when `IPeekResultFactory` is not available. The type does not function correctly without it.

**Comment by sharwell** [comment]
@KirillOsenkov Note that we depend heavily on graceful degradation of the MEF catalog when imports are omitted. Much of the restricted IVT layers are implemented on this basis, so it will not be possible to compose Roslyn using a "strict" mechanism that expects all required imports to be satisfied.

File: src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs (hunk 2 above)
Comment (KirillOsenkov) [reply src/EditorFeatures/Core/Implementation/Peek/PeekableItemSourceProvider.cs]: I'm making the type work properly without it below, by just returning null if the part is null.

**Comment by KirillOsenkov** [comment]
VSMac relies heavily on inspecting MEF composition errors to find actual issues. This is a very valuable tool for us as we live in a world where we insert a patchwork of dependencies and we need to make sure they all work together well. So far we have been successful in maintaining error-free composition and this is the only issue.

Silent rejection terrifies me as large components of the graph can disappear and then we spend time investigating why a feature is missing or not behaving as intended. Those scare me way more than a hypothetical null-ref in a tiny class that will probably never change.

I'd like to continue to disable silent rejection in the VSMac MEF composition and I humbly ask the Roslyn team to allow this one little issue. AllowDefault is a pattern widely used already and I prefer defending against optional imports being null to investigating parts inexplicably missing from composition.

**Comment by KirillOsenkov** [comment]
@sharwell and I have been discussing this offline quite extensively and I want to make sure I get more input from other involved team members. @dotnet/roslyn-ide 

I have to confess that the design approach of using part rejection to express exports that are only present when another component is present is new to me. I have heard of rejection since 2009 original MEF but it was always considered a huge source of unforeseen errors as rejection tends to cascade and large portions of the graph can just disappear without a trace and good luck debugging why is my type not loading. 

This is why the original MEF has introduced the DisableSilentRejection flag and we've been trained to set it always to have explicit errors when parts were not loaded because of missing required imports. In my experience we have spent a lot of time investigating issues where a part was not loaded and turning errors on showed us why.

It looks like the Roslyn team has made a decision to deliberately introduce errors into the VSMEF composition by design as part of expressing "don't export this part unless someone will be actually importing this subsystem". I understand that the VS MEF composition may already contain lots of errors ("broken windows") but to my knowledge this is the first time someone is making more errors in the composition and its intentional. 

I'm curious what is the proposal to debug real MEF failures when a type isn't loaded because an import is missing unintentially due to a real bug. I'm also curious if there could be other MEF patterns that would let you express this without introducing errors into the composition.

Additionally I still don't quite understand why this PR is a problem and what real issues will we have if we merge it in its current form. To me AllowDefault imports are an established MEF pattern widely used in Visual Studio and it clearly expresses that an import is optional in some hosts.

@sandyarmstrong @CyrusNajmabadi @tmat @jinujoseph

**Comment by KirillOsenkov** [comment]
Also @olegtk

**Comment by CyrusNajmabadi** [comment]
> I have to confess that the design approach of using part rejection to express exports that are only present when another component is present is new to me. I have heard of rejection since 2009 original MEF but it was always considered a huge source of unforeseen errors as rejection tends to cascade and large portions of the graph can just disappear without a trace and good luck debugging why is my type not loading.

I strongly resonate with that.  The number of times i've had mef composition failures without being able to figure out what was going on is in the triple digits.  I remember losing hair over how awful this was.

If the situation still sucks, and part rejection is still an opaque and painful thing to figure out, i would opt toward avoiding rejection semantics.  If, however, part rejection is very easy to understand now, i could feel differently.

**Comment by KirillOsenkov** [comment]
I filed a design suggestion on vs-mef to formalize the rejection semantics: 
https://github.com/microsoft/vs-mef/issues/178

This way you have to explicitly opt-in to rejection in a place where you expect it. Unforeseen rejections caused by actual bugs will still cause errors in the composition.

**Comment by sharwell** [comment]
@jasonmalinowski and I have proposed that VS for Mac exclude the type from the catalog prior to finalizing the composition. This allows Roslyn to continue operating under the current design, and allows VS for Mac to work around the issue by excluding one or more specific types to avoid composition errors. Here is an example of a situation where parts are excluded from a MEF catalog:

https://github.com/dotnet/roslyn/blob/9bca5a2861d3ebb2f34bc36536d96d478898c64d/src/Workspaces/CoreTestUtilities/MEF/ExportProviderCache.cs#L137-L144

**Comment by sharwell** [comment]
> If the situation still sucks, and part rejection is still an opaque and painful thing to figure out

Part rejection is easy to figure out. You can get a feel for it by stepping through this block:

https://github.com/dotnet/roslyn/blob/9bca5a2861d3ebb2f34bc36536d96d478898c64d/src/Workspaces/CoreTestUtilities/MEF/ExportProviderCache.cs#L211-L238

**Comment by olegtk** [comment]
@AArnott

**Comment by jasonmalinowski** [comment]
Yeah I absolutely agree that part rejection can be a royal pain when you're trying to debug something, but at the same time it's a tool that exists and is often useful. In this _particular_ PR we're fairly lucky in the API can easily be implemented to no-op if a dependent part exists but that solution isn't always usable. Sometimes trying to deal with a missing export is complicated, and in some sense AllowDefault might mean your composition is free from errors, but isn't in a remotely tested state.

It's also a bit of a layering violation if we're having to make changes to work around the lack of support for something for one host. We can do code to mutate VS for Mac's composition to say things like "we want to throw out parts that consume this known missing part", and it seems like that instead moves the declaration of "what might be missing" to the host that knows what it doesn't support. When the support comes online, then you can update the code in VS for Mac and not have to go to other repos to remove the workarounds.

:+1: for trying to make this a bit more declarative so we can have both sides: part rejection when it's useful without losing diagnosability.

**Comment by AArnott** [comment]
Several points to add to the discussion:

1. Rejection in MEF is a first class feature and it's fine to depend on it.
1. Rejection is a multi-pass process during composition, and each time we compose and have to reject a part, it requires restarting the composition, which can take seconds on some machines. So *adding* rejection to the common case may slow down VS setup or cache-invalidated VS launches by several seconds.
1. The diagnostic log is indeed vital to many people's investigation, so adding 'by design' rejection will certainly add noise and complicate the investigation.
1. CPS used to depend *heavily* on rejection but when we switched from .NET MEF to VS MEF we entirely switched away from that. Our goal was different from yours though, but effectively we do just deal with conditional imports. In our case we have *n* exports so all of these conditional imports are collections.
1. `AllowDefault=true` is another first class feature of MEF and is usually a better choice when the export may not be present. Rejection uniquely can take out entire transitive chains of exports if the leaf export is missing. 

**In summary**

If AllowDefault=true is sufficient for you and all you have to do is a null check before using it, I would recommend you take that route.

**Comment by KirillOsenkov** [comment]
FYI @Therzok we may need to start filtering individual types from the composition.

**Comment by tmat** [comment]
@sharwell @jasonmalinowski Given @AArnott's recommendations, what's the verdict for Roslyn? 
Applies to https://github.com/dotnet/roslyn/pull/42144 also.

**Comment by sharwell** [comment]
My recommendation is to continue with part rejection. It's the only way to keep IVT isolation and still continue to behave as though the partner team provided the original part.

Or in other words, we have a slightly more expensive setup operation (~one time) in exchange for a more accurate and efficient runtime configuration.

**Comment by KirillOsenkov** [comment]
I'm OK with this. Feel free to close the PR.
